core/chain.py

```
@@ -1,22 +1,141 @@
+                
+                stored_merkle = curr_data.get("merkle_root")
+                computed_merkle = curr_block.merkle_root
+                
+                if stored_merkle != computed_merkle:
+                    logger.warning(f"Loaded chain has invalid merkle_root at block {i}. Rejecting loaded chain.")
+                    self.chain = []
+                    break
+            else:
+                if data.get("state"):
+                    self.state = State.from_dict(data["state"])
+                    logger.info(f"Loaded chain with {len(self.chain)} blocks from {self._chain_file}")
```

Ruff (0.15.1)

[warning] Logging statement uses f-string

(G004)